service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/CaseExtensionController.java

```diff
@@ -0,0 +1,102 @@
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStartEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        caseCompletionDate = getCaseCompletionDate(caseData);
+
+        caseDetails.getData().put("shouldBeCompletedByDate", formatLocalDateToString(caseCompletionDate, DATE));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackrequest) {
+        CaseDetails caseDetails = callbackrequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        eightWeekExtensionDate = getCaseCompletionDateFor8WeekExtension(caseData);
+
+        caseDetails.getData().put("extensionDateEightWeeks", formatLocalDateToString(eightWeekExtensionDate,
+                DATE));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .errors(validateGroupService.validateGroup(caseData, CaseExtensionGroup.class))
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+
+        caseCompletionDate = getCaseCompletionDate(caseDetails);
+        caseDetails.getData().put("caseCompletionDate", caseCompletionDate);
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    private LocalDate getCaseCompletionDate(CaseDetails caseDetails) {
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        if (caseDetails.getData().get("caseExtensionTimeList").equals("EightWeekExtension")) {
+            if (caseDetails.getData().get("caseExtensionTimeConfirmationList").equals("EightWeekExtension")) {
+                return getCaseCompletionDateFor8WeekExtension(caseData);
+            } else {
+                return caseData.getEightWeeksExtensionDateOther();
+            }
+        } else {
+            return caseData.getExtensionDateOther();
+        }
+    }
+
+    private LocalDate getCaseCompletionDateFor8WeekExtension(CaseData caseData) {
+        return getCaseCompletionDate(caseData).plusWeeks(8);
+    }
+
+    private LocalDate getCaseCompletionDate(CaseData caseData) {
+        if (isEmpty(caseData.getCaseCompletionDate())) {
+            return caseData.getDateSubmitted();
+        } else {
+            return caseData.getCaseCompletionDate();
+        }
+    }
```

Comment:
Would this be useful as a method on caseData itself?

Reply:
Moving to caseExtensionService as discussed